pkg/proxy/iptables/proxier.go

```diff
@@ -1225,7 +1225,21 @@ func (proxier *Proxier) syncProxyRules() {
 		// metrics come out right, so we just compute them and throw them away.
 		if tryPartialSync {
 			svcNN := svcName.NamespacedName.String()
-			if !serviceChanged.Has(svcNN) && !endpointsChanged.Has(svcNN) {
+			canSkip := !serviceChanged.Has(svcNN) && !endpointsChanged.Has(svcNN)
+			if canSkip && len(existingNATChains) > 0 {
```

Comment:
This will be > 0 when

> 	if !proxier.largeClusterMode || time.Since(proxier.lastIPTablesCleanup) > proxier.syncPeriod {

that means it can be running a lot of times

I want to be cautious here,  we only have one report of this problem, the problem is not easy to reproduce, we couldn't discard other environmental factors, we don't know if there are more actors doing iptables stuff on that environment, and also the fact that a higher version of kube-proxy without any related changes behave differently (better) is another question hard to answer and that may indicate is not a bug here

@juliantaylor  will you be fine if we give you a custom build of kube-proxy with this logging to root cause the behavior you are finding in https://github.com/kubernetes/kubernetes/issues/121362#issuecomment-1808499263

Reply:
In most clusters, it will run on every sync. But it's not doing a lot of work...

>  we only have one report of this problem

Yeah, I guess I'm worried that we only have one person who has _noticed_ it, not that we only have one person who is experiencing it.

We could merge just the extra logging stuff and _not_ merge the attempted recovery if you want...

Reply:
+1 to the extra logging

 > not merge the attempted recovery if you want...

if you can change this by a metric that will awesome, because this way I can get signal and identify more clusters

Reply:
I can maybe try to test a version with more logging, a problem is though that we are in the process of replacing kube-proxy with a fully calico based solution (instead of kube-proxy + calico) but we can still have kube-proxy running along side it with no effect beside writing unused iptables which I could test, but I can't promise to do this in a timely manner.

As for there only being one report is not that surprising, a practical impact on externalTrafficPolicy Cluster services is very unlikely, as it manifests only as some missing endpoints on some of the clusters nodes. Only with Local policy can you properly observe it and only when the problem hits those services and your monitoring sees the outage from only a subset of your local serving nodes.

I am wondering has anyone tried checking if it can be seen in other environments?
I have updated the scripts in gist of the issue used to see the problem to include clusterip services and exclude stuff based on `endpoints.kubernetes.io/last-change-trigger-time` in the mean time.

Reply:
> As for there only being one report is not that surprising, a practical impact on externalTrafficPolicy Cluster services is very unlikely

we can not backport a change like this just to debug a cluster

> instead of kube-proxy + calico

you are running calico in this cluster too?
when there are multiple actors in a system using iptables these things can happen, we don't know 100% if this is a problem in kube-proxy alone or if there is something else conflicting with kube-proxy, do we?

Reply:
We don't really know a lot, though FWIW, unless there was a process that was actively removing kube-proxy's rules, I'm not sure how other actors could cause this problem either.

Well, OK, so like, if you had something doing "iptables-save | (make some changes) | iptables-restore", then that could cause race conditions where the other component would end up undoing changes kube-proxy made in between when it had done the "iptables-save" and "iptables-restore". (Like https://github.com/cloudnativelabs/kube-router/issues/1370, though the problem here wouldn't actually be kube-router, since (a) they fixed that code, and (b) they only did that on the `filter` table, not `nat`, which is where we're seeing the problem.) I'm pretty sure Calico does not do this. I'm sure we would have run into problems long ago if they did.

Reply:
Actually, the fact that there is evidence of both occasional missing service rules _and_ occasional missing endpoint rules in @juliantaylor 's cluster is pretty odd (in terms of possible ways kube-proxy could be screwing up), and may be a hint that the problem is "someone is deleting rules after kube-proxy creates them" rather than "kube-proxy is failing to create rules"...

Reply:
> Actually, the fact that there is evidence of both occasional missing service rules _and_ occasional missing endpoint rules in @juliantaylor 's cluster is pretty odd (in terms of possible ways kube-proxy could be screwing up), and may be a hint that the problem is "someone is deleting rules after kube-proxy creates them" rather than "kube-proxy is failing to create rules"...

We have never seen missing service rules?
Unless you refer to the missing local drop rule when no local endpoints exist. This is caused the code always writing service rules based on all endpoints but only adds the endpoint change when it thinks the endpoints of specific service have changed which is probably the thing being missed.
For me this behaviour implies an issue in kube-proxy and not an external component.

Also calico is not exactly rarely used, I'd expect you still want to know whats up here.

Reply:
> We have never seen missing service rules?

You haven't _noticed_ them, because they immediately cause syncs to fail and force a full resync. But there was one in the logs you provided: https://github.com/kubernetes/kubernetes/issues/121362#issuecomment-1777551011